### ISSUE TITLE
How to visualize the attention map?

### ISSUE THREAD

**Issue by youngwanLEE** [guoxih/region-based-non-local-network#3]
Hi,

I'm very impressed by your excellent work.

I wonder how to visualize [the attention maps](https://github.com/guoxih/region-based-non-local-network/blob/main/visualization.jpg).

Thanks

**Comment by edhuang1** [comment]
Hi, 

Thank you for interesting my work.

As defined in [GCNet: Non-local Networks Meet Squeeze-Excitation Networks and Beyond](https://arxiv.org/abs/1904.11492),  the outer product of a feature map and itself presents the pair-wise relationship of any two postions in the feature map. 
The outpter product is of shape THW x THW. So when we secify a point on its first dimension, we can plot out the attention map for that point.

Below is the code I use to visualize the attention maps, but I did not manage well on it, so I cannot guarantee it works:

    from ops.rnl import _NonLocalBlockND
    subsampling = False
    count = 0
    for m in model.modules():
        if isinstance(m, _NonLocalBlockND):
            count += 1
            if count ==2:
                print(m.attention_maps.shape)
                print(m.attention_size)
                break
    T, H, W = m.attention_size[2], m.attention_size[3], m.attention_size[4]

    if subsampling:
        T2, H2, W2 = T, H//2, W//2
    else:
        T2, H2, W2 = T, H, W
    attention = m.attention_maps.reshape([m.attention_maps.size(0), T, H, W, T2, H2, W2]).detach().cpu().numpy()
    print('shape of attention: {}'.format(attention.shape))

    import matplotlib.pyplot as plt
    def plot_attention(query_position):
        norm_fn = lambda x: (x - np.min(x, axis=None, keepdims=True))/(np.max(x, axis=None, keepdims=True)-np.min(x, axis=None, keepdims=True))
        _BATCH_SIZE = attention.shape[0]
        classmap_vis_v2 = attention[:,query_position[0],query_position[1],query_position[2]]
        point_map = np.zeros((_BATCH_SIZE,T2,H2,W2))
        point_map[:,query_position[0],query_position[1],query_position[2]] = 1
        point_map = point_map.reshape((_BATCH_SIZE,T2,H2,W2))

        from scipy.ndimage import zoom
        import scipy
        from PIL import Image

        resized_point_map = zoom(point_map, (1, 1, 224//H2, 224/W2), order=0)
        resized_point_map = np.concatenate([resized_point_map[:,_i] for _i in range(T2)], axis=-1)
        masked = np.ma.masked_where(resized_point_map == 0, resized_point_map)


        current_images = input.detach().cpu().numpy()

        current_images = current_images.reshape((-1,8,3,224,224)).transpose((0,1,3,4,2))

        resized_classmap_vis_v2 = []
        for i in range(_BATCH_SIZE):
            item = [np.array(Image.fromarray(_img).resize([224,224], Image.ANTIALIAS)) for _img in classmap_vis_v2[i]]
            resized_classmap_vis_v2.append(item)
        resized_classmap_vis_v2 = np.array(resized_classmap_vis_v2)
        resized_classmap_vis_v2 = np.concatenate([resized_classmap_vis_v2[:,_i] for _i in range(T2)], axis=-1)

        current_images = norm_fn(current_images)


        current_images = np.concatenate([current_images[:,_i] for _i in range(T2)], axis=-2)
        fig = plt.figure(figsize=(25, 15))
        for i in range(_BATCH_SIZE):
            if i == 4:
                plt.subplot(_BATCH_SIZE, 1, i+1)
                plt.imshow(current_images[i])
                plt.imshow(resized_classmap_vis_v2[i], cmap=plt.cm.jet, alpha=0.35, interpolation='nearest')
                plt.imshow(masked[i], alpha=1,cmap = 'RdYlBu')
                plt.xticks([]), plt.yticks([])
        fig.tight_layout()
        plt.show()

**Comment by youngwanLEE** [comment]
@guoxih 
Thanks for your reply. 👍
